src/modules/expressions.js

```diff
@@ -0,0 +1,114 @@
+import i18n from '../locales/index.js'
+
+export const EXPRESSION_TYPE_NUMBER = 'EXPRESSION_TYPE_NUMBER'
+export const EXPRESSION_TYPE_OPERATOR = 'EXPRESSION_TYPE_OPERATOR'
+export const EXPRESSION_TYPE_DATA = 'EXPRESSION_TYPE_DATA'
+
+export const VALID_EXPRESSION = 'OK'
+export const INVALID_EXPRESSION = 'ERROR'
+
+export const getOperators = () => [
+    { value: '+', label: '+', type: EXPRESSION_TYPE_OPERATOR },
+    { value: '-', label: '-', type: EXPRESSION_TYPE_OPERATOR },
+    { value: '*', label: '×', type: EXPRESSION_TYPE_OPERATOR },
+    { value: '/', label: '/', type: EXPRESSION_TYPE_OPERATOR },
+    { value: '(', label: '(', type: EXPRESSION_TYPE_OPERATOR },
+    { value: ')', label: ')', type: EXPRESSION_TYPE_OPERATOR },
+    {
+        value: '',
+        label: i18n.t('Number'),
+        type: EXPRESSION_TYPE_NUMBER,
+    },
+]
+
+export const parseExpression = (input) => {
+    const regex = /(#{[a-zA-Z0-9#.]+}|[+\-*/()])|(\d+(\.\d+)?)/g
+    return input.match(regex) || []
+}
+
+export const parseExpressionToArray = (expression = '', metadata = []) => {
+    return (
+        parseExpression(expression).map((value) => {
+            if (value.startsWith('#{') && value.endsWith('}')) {
+                const id = value.slice(2, -1)
+                const label =
+                    metadata.find((item) => item.id === id)?.name || id
+                return { value, label, type: EXPRESSION_TYPE_DATA }
+            }
+
+            if (isNaN(value)) {
+                return {
+                    value,
+                    label: getOperators().find((op) => op.value === value)
+                        .label,
+                    type: EXPRESSION_TYPE_OPERATOR,
+                }
+            }
+
+            return {
+                value,
+                label: value,
+                type: EXPRESSION_TYPE_NUMBER,
+            }
+        }) || []
+    )
+}
+
+export const parseArrayToExpression = (input = []) =>
+    input.map((item) => item.value).join('')
+
+export const validateExpression = (expression) => {
+    let result
+    // TODO: two numbers next to each other
```

Comment:
I don't think we can handle this easily, as the `expression` here is already parsed to a string.